processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,743 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import com.google.common.collect.ImmutableList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import org.apache.druid.math.expr.Expr;
+import org.apache.druid.query.dimension.DimensionSpec;
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.query.filter.InDimFilter;
+import org.apache.druid.query.filter.ValueMatcher;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.DimensionSelector;
+import org.apache.druid.segment.NilColumnValueSelector;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.segment.filter.AndFilter;
+import org.apache.druid.segment.filter.Filters;
+import org.apache.druid.segment.filter.InFilter;
+import org.apache.druid.segment.filter.OrFilter;
+import org.apache.druid.segment.filter.SelectorFilter;
+import org.apache.druid.segment.join.lookup.LookupColumnSelectorFactory;
+import org.apache.druid.segment.join.lookup.LookupJoinable;
+import org.apache.druid.segment.join.table.IndexedTable;
+import org.apache.druid.segment.join.table.IndexedTableJoinable;
+import org.apache.druid.segment.virtual.ExpressionVirtualColumn;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+public class JoinFilterAnalyzer
+{
+  private static final String PUSH_DOWN_VIRTUAL_COLUMN_NAME_BASE = "JOIN-FILTER-PUSHDOWN-VIRTUAL-COLUMN-";
+  private static final ColumnSelectorFactory ALL_NULL_COLUMN_SELECTOR_FACTORY = new AllNullColumnSelectorFactory();
+
+  public static JoinFilterSplit splitFilter(
+      HashJoinSegmentStorageAdapter hashJoinSegmentStorageAdapter,
+      @Nullable Filter originalFilter
+  )
+  {
+    if (originalFilter == null) {
+      return new JoinFilterAnalyzer.JoinFilterSplit(
+          null,
+          null,
+          ImmutableList.of()
+      );
+    }
+
+    Filter normalizedFilter = Filters.convertToCNF(originalFilter);
+
+    // build the prefix and equicondition maps
+    Map<String, Expr> equiconditions = new HashMap<>();
+    Map<String, JoinableClause> prefixes = new HashMap<>();
+    for (JoinableClause clause : hashJoinSegmentStorageAdapter.getClauses()) {
+      prefixes.put(clause.getPrefix(), clause);
```

Comment:
Hm, that sounds good, but can we address this in a follow-on PR since it's not specific to the filtering code?

Reply:
I think it is specific to the filtering code, because it's handled consistently elsewhere. (The clauses check in order whether they match and the first one wins.) So this patch is introducing an inconsistency in behavior that didn't used to exist. But I do think it's fine to do in a follow on. Please raise a github issue and add a comment pointing to it.

Reply:
I opened https://github.com/apache/druid/issues/9329 and linked it in comment here